File: packages/shared/src/experimental/conversation-client.ts
import {
  serverHasCapability,
  SERVER_CAPABILITIES,
  type ServerCapabilitySource,
} from "../server-capabilities.js";
import { CONVERSATION_API_REVISION } from "./conversation-protocol.js";
import {
  ConversationQuerySchema,
  IdSchema,
  decodeSnapshot,
  type ConversationQuery,
  type SnapshotEnvelope,
} from "./simple-client.generated.js";

export type ConversationAvailability =
  | "available"
  | "update-required"
  | "revision-mismatch";
export function conversationAvailability(
  version: ServerCapabilitySource & {
    experimentalSimpleClientApiRevision?: unknown;
  },
): ConversationAvailability {
  if (
    !serverHasCapability(
      version,
      SERVER_CAPABILITIES.experimentalConversation,
    ) ||
    typeof version.experimentalSimpleClientApiRevision !== "string"
  )
    return "update-required";
  return version.experimentalSimpleClientApiRevision ===
    CONVERSATION_API_REVISION
    ? "available"
    : "revision-mismatch";
}

/** A binding belongs to exactly one client-owned source, including ID collisions. */
export class ConversationBinding {
  private sequence = -1;
  private closed = false;
  constructor(
    readonly sourceId: string,
    readonly subscriptionId: string,
    readonly sessionId: string,
  ) {
    IdSchema.parse(subscriptionId);
    IdSchema.parse(sessionId);
    if (!sourceId) throw new Error("Conversation requires a source binding");
  }
  close(): void {
    this.closed = true;
  }
  accept(sourceId: string, encoded: string): SnapshotEnvelope | null {
    if (this.closed || sourceId !== this.sourceId) return null;
    const snapshot = decodeSnapshot(encoded);
    if (
      snapshot.subscriptionId !== this.subscriptionId ||
      snapshot.sequence <= this.sequence
    )
      return null;
    if (
      snapshot.view.kind === "conversation" &&
      snapshot.view.sessionId !== this.sessionId
    )
      throw new Error("Conversation session mismatch");
    this.sequence = snapshot.sequence;
    return snapshot;
  }
}

export function conversationReadPath(
  query: ConversationQuery,
  subscriptionId: string,
): string {
  ConversationQuerySchema.parse(query);
  IdSchema.parse(subscriptionId);
  const params = new URLSearchParams({
    apiRevision: CONVERSATION_API_REVISION,
    subscriptionId,
    sessionId: query.sessionId,
    maxMessages: String(query.maxMessages),
  });
  if (query.anchorMessageId !== null)
    params.set("anchorMessageId", query.anchorMessageId);
  return `/api/experimental/conversation?${params}`;
}

/** Gate before sending any experimental operation; transport/auth remain owned by the caller. */
export async function readConversation(options: {
  sourceId: string;
  subscriptionId: string;
  query: ConversationQuery;
  version: ServerCapabilitySource & {
    experimentalSimpleClientApiRevision?: unknown;
  };
  request: (path: string) => Promise<string>;
}): Promise<
  | { sourceId: string; snapshot: SnapshotEnvelope }
  | {
      sourceId: string;
      unavailable: Exclude<ConversationAvailability, "available">;
    }
> {
  const availability = conversationAvailability(options.version);
  if (availability !== "available")
    return { sourceId: options.sourceId, unavailable: availability };
  const binding = new ConversationBinding(
    options.sourceId,
    options.subscriptionId,
    options.query.sessionId,
  );
  const snapshot = binding.accept(
    options.sourceId,
    await options.request(
      conversationReadPath(options.query, options.subscriptionId),
    ),
  );
  if (snapshot?.sequence !== 0)
    throw new Error("Invalid initial Conversation binding");
  return { sourceId: options.sourceId, snapshot };
}

File: packages/android/app/src/main/java/com/yepanywhere/mobile/experimental/ConversationClient.kt
package com.yepanywhere.mobile.experimental

import com.yepanywhere.mobile.connection.YaApiException
import com.yepanywhere.mobile.connection.YaApiResponse
import com.yepanywhere.mobile.connection.YaConnectionLease
import java.net.URLEncoder
import java.util.UUID
import org.json.JSONObject

const val CONVERSATION_API_REVISION = "simple-client-spike-1"
private const val CAPABILITY = "experimental-simple-client-conversation"

enum class ConversationAvailability { AVAILABLE, UPDATE_REQUIRED, REVISION_MISMATCH }

fun conversationAvailability(version: JSONObject): ConversationAvailability {
    fun contains(field: String): Boolean {
        val entries = version.optJSONArray(field) ?: return false
        return (0 until entries.length()).any { entries.opt(it) == CAPABILITY }
    }
    fun hasBit(field: String): Boolean {
        val words = version.optJSONArray(field) ?: return false
        return (0 until words.length()).any {
            val pair = words.optJSONArray(it)
            val index = pair?.opt(0) as? Number
            val bits = pair?.opt(1) as? Number
            index?.toDouble() == 2.0 && bits != null &&
                bits.toDouble() == bits.toLong().toDouble() && bits.toLong() in 0..0xffff_ffffL &&
                bits.toLong() and 32L != 0L
        }
    }
    if (version.opt("experimentalSimpleClientApiRevision") !is String) return ConversationAvailability.UPDATE_REQUIRED
    if (!contains("capabilities") && !contains("capabilityExtensions") && !hasBit("capabilityBits") && !hasBit("optionalCapabilityBits"))
        return ConversationAvailability.UPDATE_REQUIRED
    return if (version.opt("experimentalSimpleClientApiRevision") == CONVERSATION_API_REVISION)
        ConversationAvailability.AVAILABLE else ConversationAvailability.REVISION_MISMATCH
}

class ConversationBinding(val sourceId: String, val subscriptionId: String, val sessionId: String) {
    private var sequence = -1
    private var closed = false
    init {
        require(sourceId.isNotEmpty())
        SimpleClientContract.decodeId(subscriptionId)
        SimpleClientContract.decodeId(sessionId)
    }
    fun close() { closed = true }
    fun accept(sourceId: String, encoded: String): SnapshotEnvelope? {
        if (closed || sourceId != this.sourceId) return null
        val snapshot = SimpleClientContract.decodeSnapshot(encoded)
        if (snapshot.subscriptionId != subscriptionId || snapshot.sequence <= sequence) return null
        val view = snapshot.view
        require(view !is Conversation || view.sessionId == sessionId) { "Conversation session mismatch" }
        sequence = snapshot.sequence
        return snapshot
    }
}

sealed interface ConversationReadResult {
    val sourceId: String
    data class Snapshot(override val sourceId: String, val snapshot: SnapshotEnvelope) : ConversationReadResult
    data class Unavailable(override val sourceId: String, val reason: ConversationAvailability) : ConversationReadResult
}

/** The caller's foreground lease owns authentication and connection lifetime. */
class ConversationClient(
    private val sourceId: String,
    private val request: suspend (String) -> YaApiResponse,
) {
    constructor(sourceId: String, lease: YaConnectionLease) : this(sourceId, { path -> lease.request("GET", path) })

    suspend fun read(query: ConversationQuery): ConversationReadResult {
        val version = request("/api/version").successfulBody() as? JSONObject
            ?: error("Invalid server version response")
        val availability = conversationAvailability(version)
        if (availability != ConversationAvailability.AVAILABLE)
            return ConversationReadResult.Unavailable(sourceId, availability)
        SimpleClientContract.decodeConversationQuery(JSONObject()
            .put("sessionId", query.sessionId).put("maxMessages", query.maxMessages)
            .put("anchorMessageId", query.anchorMessageId ?: JSONObject.NULL))
        val binding = ConversationBinding(sourceId, UUID.randomUUID().toString(), query.sessionId)
        val fields = linkedMapOf("apiRevision" to CONVERSATION_API_REVISION, "subscriptionId" to binding.subscriptionId,
            "sessionId" to query.sessionId, "maxMessages" to query.maxMessages.toString())
        query.anchorMessageId?.let { fields["anchorMessageId"] = it }
        val params = fields.entries.joinToString("&") { (key, value) -> "$key=${URLEncoder.encode(value, "UTF-8")}" }
        val body = request("/api/experimental/conversation?$params").successfulBody()
        val snapshot = binding.accept(sourceId, body.toString()) ?: error("Invalid Conversation binding")
        require(snapshot.sequence == 0) { "Invalid initial Conversation sequence" }
        return ConversationReadResult.Snapshot(sourceId, snapshot)
    }

    private fun YaApiResponse.successfulBody(): Any? {
        if (status !in 200..299) throw YaApiException(this)
        return body
    }
}
